codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -438,9 +455,176 @@ public void operationComplete(ChannelFuture future) throws Exception {
+            super(ctx, stream, padding, endOfStream, promise);
+            this.data = data;
+            size = dataLength + padding;
+        }
+
+        @Override
+        public long size() {
+            return size;
+        }
+
+        /**
+         * Return {@link Integer#MAX_VALUE} if the actual size is larger than
+         * {@link Integer#MAX_VALUE}.
+         */
+        protected abstract int writableData();
+
+        protected abstract T sliceData(int length);
+
+        protected abstract void writeData(T toWrite, int padding, boolean endStream,
+                ChannelPromise writePromise);
+
+        private ChannelPromise selectPromise(int bytesWritten) {
+            if (size == bytesWritten && !promise.isVoid()) {
+                // Can use the original promise if it's the last write
+                return promise;
+            } else {
+                // Create a new promise and listen to it for failure
+                ChannelPromise newPromise = ctx.newPromise();
+                newPromise.addListener(this);
+                return newPromise;
+            }
+        }
+
+        private int writePaddingOnly(int allowedFrameSize, int bytesWritten) {
+            int writablePadding = Math.min(allowedFrameSize, padding);
+            padding -= writablePadding;
+            bytesWritten += writablePadding;
+            ChannelPromise writePromise = selectPromise(bytesWritten);
+            frameWriter().writeData(ctx, stream.id(), Unpooled.EMPTY_BUFFER, writablePadding,
+                    size == bytesWritten && endOfStream, writePromise);
+            return bytesWritten;
+        }
+
+        @SuppressWarnings("unchecked")
+        private int writeFrame(int allowedFrameSize, int bytesWritten) {
+            T toWrite;
+            int writableData = writableData();
+            // Let data consume the frame before padding.
+            if (writableData > allowedFrameSize) {
+                writableData = allowedFrameSize;
+                toWrite = (T) sliceData(allowedFrameSize).retain();
+            } else {
+                // We're going to write the full buffer which will cause it to be released. Set data
+                // to null to indicate subsequent writes(if any) that we only need to write padding.
+                toWrite = data;
+                data = null;
+            }
+            int writablePadding = Math.min(allowedFrameSize - writableData, padding);
+            padding -= writablePadding;
+            bytesWritten += writableData + writablePadding;
+            ChannelPromise writePromise = selectPromise(bytesWritten);
+            writeData(toWrite, writablePadding, size == bytesWritten && endOfStream, writePromise);
+            return bytesWritten;
+        }
+
+        @Override
+        public void write(int allowedBytes) {
+            int bytesWritten = 0;
+            if (data == null || (allowedBytes == 0 && size != 0)) {
```

Comment:
o, it is `||`, not `&&`. Let me fix it.
I separated the logic of writing padding only and writing frame because we do not have an `EmptyFileRegion` and it makes no sense to have an `EmptyFileRegion` I think? There must be a file associate with a `FileRegion`.